jgrapht-core/src/main/java/org/jgrapht/alg/tour/WarnsdorffRule.java

```diff
@@ -0,0 +1,873 @@
+* the Eclipse Foundation.
+*/
+
+package org.jgrapht.alg.tour;
+
+import java.util.*;
+
+/*
+  Implementation of {@literal <a href = "https://en.wikipedia.org/wiki/Knight%27s_tour#Warnsdorf's_rule">}Warnsdorff's rule{@literal </a>}
+  - heuristic for finding a knight's tour on chessboards.
+
+  A knight's tour is a sequence of moves of a knight on a chessboard such that the knight visits every square only once.
+  If the knight ends on a square that is one knight's move from the beginning square
+  (so that it could tour the board again immediately, following the same path), the tour is closed, otherwise it is open.
+
+  The knight's tour problem is the mathematical problem of finding a knight's tour.
+
+  Description of the Warnsdorff's rule: set a start cell. Always proceed to the cell that have the fewest onward moves.
+  In case of a tie(i.e. there exist more than one possible choice for the next cell) go to the cell with largest Euclidean distance from the center of the board.
+
+  This implementation also allows you to find a structured knight's tour.
+
+  Knight's tour on board of size $n \times m$ is called structured if it contains the following $8$ UNDIRECTED moves:
+
+  1). $(1, 0) \to (0, 2)$.
+  2). $(2, 0) \to (0, 1)$.
+  3). $(n - 3, 0) \to (n - 1, 1)$.
+  4). $(n - 2, 0) \to (n - 1, 2)$.
+  5). $(0, m - 3) \to (1, m - 1)$.
+  6). $(0, m - 2) \to (2, m - 1)$.
+  7). $(n - 3, m - 1) \to (n - 1, m - 2)$.
+  8). $(n - 2, m - 1) \to (n - 1, m - 3)$.
+
+  #########################################
+  #*12*********************************34*#
+  #2*************************************3#
+  #1*************************************4#
+  #***************************************#
+  #***************************************#
+  #***************************************#
+  #***************************************#
+  #***************************************#
+  #***************************************#
+  #***************************************#
+  #***************************************#
+  #***************************************#
+  #***************************************#
+  #***************************************#
+  #5*************************************7#
+  #4*************************************6#
+  #*54*********************************67*#
+  #########################################
+
+  If you are confused with the formal definition of the structured knight's tour
+  please refer to illustration on the page 3 of the paper
+  "An efficient algorithm for the Knight’s tour problem " by Ian Parberry.
+
+  One more feature of this implementation is that it provides an option to return a shifted knight's tour,
+  where all cell's coordinates are shifted by some values.
+  Basically it is the same as knight's tour of some piece of the board.
+ */
+
+/**
+ * Class that represents a chessboard cell.
+ * @param <F> type of the first attribute.
+ * @param <S> type of the second attribute.
+ */
+
+class ChessBoardCell<F, S>
```

Comment:
Is there a reason/need/advantage to make this class generic? Also, this looks suspiciously much like the Pair<F,S> implementation already present in JGraphT. You could use Pair, or extend Pair.

Reply:
I was looking into it, but `Pair` class has 'final' modifiers. May I remove them?